native/src/main/java/io/ballerina/lib/data/csvdata/utils/CsvUtils.java

```diff
@@ -0,0 +1,373 @@
+                continue;
+            }
+            return Optional.of(constituentType);
+        }
+        return Optional.empty();
+    }
+
+    public static Object convertToBasicType(Object csv, Type targetType, CsvConfig config) {
+        if (csv == null) {
+            csv = config.nilValue;
+        }
+        return ValueUtils.convert(csv, targetType);
+    }
+
+    public static void checkRequiredFieldsAndLogError(Map<String, Field> filedHierarchy, boolean absentAsNilableType) {
+        filedHierarchy.values().forEach(field -> {
+            if (absentAsNilableType && field.getFieldType().isNilable()) {
+                return;
+            }
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_FIELD_IN_CSV, field.getFieldName());
+            }
+        });
+    }
+
+    public static boolean isHeaderFieldsEmpty(Map<String, Field> currentField) {
+        for (Field field: currentField.values()) {
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean checkTypeCompatibility(Type constraintType, Object csv, boolean stringConversion) {
+        int tag = constraintType.getTag();
+        if (csv instanceof BString) {
+            if (stringConversion || TypeTags.isStringTypeTag(tag) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+            return false;
+        }
+        if (csv instanceof Long) {
+            if (TypeTags.isIntegerTypeTag(tag) || tag == TypeTags.FLOAT_TAG || tag == TypeTags.DECIMAL_TAG
+                    || tag == TypeTags.BYTE_TAG || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+            return false;
+        }
+        if (csv instanceof BDecimal) {
+            if ((tag == TypeTags.DECIMAL_TAG
+                    || tag == TypeTags.FLOAT_TAG || TypeTags.isIntegerTypeTag(tag)) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        if (csv instanceof Double) {
+            if ((tag == TypeTags.FLOAT_TAG
+                    || tag == TypeTags.DECIMAL_TAG || TypeTags.isIntegerTypeTag(tag)) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+            return false;
+        }
+        if (csv instanceof Boolean) {
+            if (tag == TypeTags.BOOLEAN_TAG || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+            return false;
+        }
+        if (csv == null) {
+            return tag == TypeTags.NULL_TAG || isJsonOrAnyDataOrAny(tag);
+        }
+        return false;
+    }
+
+    private static boolean isJsonOrAnyDataOrAny(int tag) {
+        return tag == TypeTags.JSON_TAG || tag == TypeTags.ANYDATA_TAG || tag == TypeTags.ANY_TAG;
+    }
+
+    public static int getTheActualExpectedType(Type type) {
```

Comment:
Can we give this a better name it seems we are returning the expected size not type